### ISSUE TITLE
DenoJS SDK

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#4368]
## Overview

Dagger has a [NodeJS SDK](https://docs.dagger.io/sdk/nodejs). What if it also had a [DenoJS](https://deno.land) SDK?

Deno is a solid alternative to NodeJS, by the creator of NodeJS, and has great traction. It could help gain additional exposure with the kind of forward-thinking developers that tend to enjoy Dagger.

**Comment by slumbering** [comment]
That's a great idea !! 

However NodeJS is still a huge ecosystem that will live for many years. 
Deno has many advantage and has great traction indeed but the community is still pretty small compare to NodeJS.
I'm just wondering how much effort do we want to put on create a "new" SDK. I'm not familiar with Deno usage, so maybe a migration from Node to Deno could be not that painfull (but I'm also always sceptical about painless migration...).

**Comment by helderco** [comment]
Isn't Deno just another JavaScript/Typescript runtime? It should be compatible with the current SDK, at least in code. Probably no need to create a new SDK.

The "NodeJS" name is unfortunate. It supports both Typescript and JavaScript so we avoided both names and chose NodeJS. Now there's other runtimes like Deno so again it feels restrictive.

The **common denominator** here is JavaScript. We should probably rename the NodeJS SDK to just the JS SDK. Even though we favor Typescript, it's a superset of JavaScript and it's all a part of the JS ecosystem anyway.

Even if we need to split, I'd rather have a JS SDK and a TS SDK (language based instead of runtime based), where the JS SDK could be auto generated entirely from the TS one.

**Comment by TomChv** [comment]
> Dagger has a [NodeJS SDK](https://docs.dagger.io/sdk/nodejs). What if it also had a [DenoJS](https://deno.land/) SDK?

I love Deno and I think it fit well to Dagger to have a Deno SDK. Especially since its usage is closer to Go (we can create binary from Deno code etc). The community is still small, but building tools in Deno can only help too!  
However it's true that Typescript code shall works on Deno, the only difference may be dependencies installed.

I think we can [use Node package](https://deno.land/manual@v1.12.2/npm_nodejs) in Deno, but I think using Deno modules would be better.

There will not much difference between those SDK though, the DX and code will be almost the same except for external library call.

I think it does not require much work, compared to creating a RustSDK for instance but it will still need time.

As a Deno user, I'm down to work on it :D

**Comment by TomChv** [comment]
After quick research, our TS GQL client is also available in Deno : https://deno.land/x/graphql_request@v4.1.0

So basically the only part to change in Deno is our static files.
The generation shall be exactly the same except the import (defined in a static header)

**Comment by NReilingh** [comment]
Happy to hear enthusiasm for Deno support!

I did some experimentation today and was able to run a Dagger pipeline from Deno successfully, with some caveats. Here is a working example:

```typescript
// main.ts
import Client, { connect } from "https://esm.sh/@dagger.io/dagger@0.6.0";

import process from 'node:process';

connect(async (client: Client) => {
  const node = client .container()
    .from("node:16-slim")
    .withExec(["node", "-v"]);

  const version = await node.stdout();

  console.log("Hello from Dagger and Node " + version);
}, { LogOutput: process.stdout });
```

`dagger run deno run -A main.ts`

## Here is what does not work:

### `import Client, { connect } from "npm:@dagger.io/dagger@0.6.0";`

This fails with an error that may or may not be blameable on Deno's NPM compat layer -- this has an issue tracking at https://github.com/denoland/deno/issues/16708

The workaround of using https://esm.sh is described in [npm via CDNs | Manual | Deno](https://deno.com/manual@v1.34.1/node/cdns).

### `deno run -A main.ts`

When invoked outside of `dagger run`, the SDK attempts to download and install the CLI, which fails with a different error:

```
error: Uncaught InitEngineSessionBinaryError: failed to download dagger cli binary: TypeError: d.pipe is not a function
    at G.<anonymous> (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:8:23837)
    at Generator.throw (<anonymous>)
    at T (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:8:22256)
    at eventLoopTick (ext:core/01_core.js:181:11)
Caused by: TypeError: d.pipe is not a function
    at https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:9:679
    at new Promise (<anonymous>)
    at G.<anonymous> (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:9:618)
    at Generator.next (<anonymous>)
    at l (https://esm.sh/v124/@dagger.io/dagger@0.6.0/deno/dagger.mjs:8:22205)
    at eventLoopTick (ext:core/01_core.js:181:11)
```

I don't know what the most elegant solution is here, but since there isn't much in the SDK's API that is Node-specific as far as I'm aware, I would suspect that this only needs to be one package (instead of two separate packages for Node.js and Deno). Conforming to Deno's Node.js compatibility layer might be all that's required, but it would be interesting to know if the Deno project suggests any best practices for publishing cross-compatible packages. (I found this tool for going in the opposite direction -- transforming a Deno module into one compatible with Node.js: https://github.com/denoland/dnt.)

**Comment by helderco** [comment]
We could also consider Deno support "experimental" by not worrying about automatic provisioning for now as it seems like the main source of incompatibility. Meaning that you'd need to run the pipeline with `dagger run`. Soon with `dagger do`. Can everything else work, outside provisioning?

**Comment by wingyplus** [comment]
@helderco It works with `dagger run` with @NReilingh solution:

```shell
$ dagger run deno run --allow-env --allow-net main.ts
█ [14.7s] deno run --allow-env --allow-net main.ts
┃ v16.20.0
┣─╮
│ ▼ from node:16
│ █ [3.65s] resolve image config for docker.io/library/node:16
│ █ [0.01s] pull docker.io/library/node:16
│ ┣ [0.01s] resolve docker.io/library/node:16@sha256:07ac708c46cc7136db9d54749c541ba1dc516dc316e126a94136083b842c2db9
│ ┣ [0.31s] ███████████████████ sha256:18e333b196442e37a54a8739afbdea755487968048a217eeb5799034d1acb6e5
│ ┣ [1.24s] ███████████████████ sha256:ffa2b6be34f250e60fc6f8ad29553caaae8e17f22e4365515c6cb9a1a68ebedf
│ ┣ [0.35s] ███████████████████ sha256:39306f6d5fdc94a315748f96c707d29011dd4dc18e0b370e38bfeb16ca1244d5
│ ┣ [0.73s] ███████████████████ sha256:a424cb15f169a0bb55bf5bfea03fc5d709a62dbd38330759e39389d316ca7476
│ ┣ [4.86s] ███████████████████ sha256:1879ae1722a31ce8aedfe4a10df652902581eab286fbbdd253684bf6ceb94cce
│ ┣ [2.78s] ███████████████████ sha256:f104dc4fba2c93f994673ea6c81b78ad44ef6b4d6984576abf9be7cf0dde960b
│ ┣ [1.21s] ███████████████████ sha256:e7349847b3241e237d7a69174fea45593592e320324afdbae5c685b0a96e8181
│ ┣ [1.66s] ███████████████████ sha256:e8371d57f7426517aead21bff5af0cf321625cac166c86214c439fb67db84243
│ ┣ [1.39s] extracting sha256:e8371d57f7426517aead21bff5af0cf321625cac166c86214c439fb67db84243
│ ┣ [0.38s] extracting sha256:e7349847b3241e237d7a69174fea45593592e320324afdbae5c685b0a96e8181
│ ┣ [1.23s] extracting sha256:f104dc4fba2c93f994673ea6c81b78ad44ef6b4d6984576abf9be7cf0dde960b
│ ┣ [3.16s] extracting sha256:1879ae1722a31ce8aedfe4a10df652902581eab286fbbdd253684bf6ceb94cce
│ ┣ [0.01s] extracting sha256:39306f6d5fdc94a315748f96c707d29011dd4dc18e0b370e38bfeb16ca1244d5
│ ┣ [0.72s] extracting sha256:ffa2b6be34f250e60fc6f8ad29553caaae8e17f22e4365515c6cb9a1a68ebedf
│ ┣ [0.05s] extracting sha256:a424cb15f169a0bb55bf5bfea03fc5d709a62dbd38330759e39389d316ca7476
│ ┣ [0.01s] extracting sha256:18e333b196442e37a54a8739afbdea755487968048a217eeb5799034d1acb6e5
│ ┣─╮ pull docker.io/library/node:16
│ ┻ │
█◀──╯ [0.96s] exec docker-entrypoint.sh node -v
┃     v16.20.0
```

NOTE: Try with this code:

```typescript
import Client, { connect } from 'https://esm.sh/@dagger.io/dagger@0.6.0';


connect(async (client: Client) => {
  const output = await client.container().from("node:16").withExec(["node", "-v"]).stdout();
  console.log(output);
});
```

**Comment by tsirysndr** [comment]
I've done some experimentation to improve the use of Deno with Dagger here [fluentci](https://github.com/fluent-ci-templates/fluentci) , so that you can run a deno module directly from https://deno.land/x/ , for example you can run `dagger run fluentci "your_deno_module"` which will be translated into `dagger run deno run --import-map=https://deno.land/x/your_deno_module/import_map.json https://deno.land/x/your_deno_module/src/dagger/runner.ts` or just `fluentci "your_deno_module"`.

**Comment by NReilingh** [comment]
New discoveries here -- I've noticed that the 0.9.7 TypeScript SDK breaks with the previously-working Deno approach. The error thrown is:

```
error: Uncaught (in promise) TypeError: The URL must be of scheme file
    at fileURLToPath (node:url:1160:11)
    at https://esm.sh/v135/@dagger.io/dagger@0.9.7/denonext/dagger.mjs:14:44486
```

And from what I can tell from the compiled and minified bundle, the call site looks something like this:

```javascript
import { fileURLToPath } from "node:url";
fileURLToPath(import.meta.url);
```

It's unclear at the moment where the root cause is, since according to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import.meta), `import.meta.url` should be starting with a `file://` scheme. Perhaps Deno breaks this. I haven't been able to narrow down which commit has the breaking change yet.

Anyhow, here's the most recent working version I've been able to come up with:

```typescript
// main.ts
import Client, { connect } from "https://esm.sh/@dagger.io/dagger@0.9.6";

import process from 'node:process';

connect(async (client: Client) => {
  const node = client .container()
    .from("node:16-slim")
    .withExec(["node", "-v"]);

  const version = await node.stdout();

  console.log("Hello from Dagger and Node " + version);
}, { LogOutput: process.stdout });
```

`dagger run deno run -A main.ts`

**Comment by slumbering** [comment]
Hey @NReilingh 👋,

I just try your `main.ts` script with the latest engine version: `v0.11.1` and Deno `v1.42.4`

```bash
$ dagger version
dagger v0.11.1 (registry.dagger.io/engine) darwin/arm64
```

package.json
```json
"dependencies": {
    "@dagger.io/dagger": "^0.11.1",
    ...
 }
```

```bash
$ dagger run deno run -A deploy.ts
Hello from Dagger and Node v16.20.2
```

Everything seems to work on my side 👍

**Comment by NReilingh** [comment]
@slumbering I think this is consistent with my last post -- to clarify, the code block above references SDK 0.9.6 in the import line, which works (on the current engine and Deno versions), but if you bump this to SDK version 0.9.7 (or later versions, _including_ 0.11.1), you get the error I described.

Your `package.json` version above will not affect the version being imported by Deno, which in my `main.ts` example is using a fully-specified import URL.

**Comment by slumbering** [comment]
> ```ts
> https://esm.sh/@dagger.io/dagger@0.9.6
> ```

Sorry about the confusion, you're right...
I can repro now.

**Comment by shykes** [comment]
We now have a Typescript SDK that is not specific to one particular runtime. Not sure if we are still considering this @TomChv, or if it's out of date?

**Comment by michaelmass** [comment]
I've tried the latest `dagger` release `0.12.3` with `deno`, but unfortunately, I get the error message `Cannot assign to read-only property 'name' of function`. Is the `deno` runtime still being considered supported by `dagger` ?

This is the command I use to run the pipeline:
``` sh
dagger run deno run -A mod.ts
```

Here's the pipeline code:
``` ts
import { connect } from "https://esm.sh/@dagger.io/dagger@0.12.3";

await connect(async (client) => {
  await client
    .container()
    .from("alpine:latest")
    .withExec(["echo", "Hello, World!"], { skipEntrypoint: true })
    .sync();
});
```

This is the error I am getting in my terminal.
![image](https://github.com/user-attachments/assets/0df48931-52bb-415c-ad56-9145c79ff00f)

**Comment by TomChv** [comment]
> I've tried the latest dagger release 0.12.3 with deno, but unfortunately, I get the error message Cannot assign to read-only property 'name' of function. Is the deno runtime still being considered supported by dagger ?

Deno has always been a community work, dagger officialy supports Node & Bun as TS runtime ([info on the setup in our doc](https://docs.dagger.io/manuals/developer/typescript/356352/runtime/))

We did several update on the engine so it's possible that the Deno module hasn't follows with the updates.

We might officialy support it later, but it's not a priority for now.

**Comment by igl** [comment]
Slightly different on deno 1.46.1  / dagger v0.12.6 
using `import { Client, connect } from 'npm:@dagger.io/dagger'` instead of esm.sh:

```sh
$ deno run -A deploy.ts
error: Uncaught SyntaxError: The requested module 'graphql-tag' does not provide an export named 'gql'
    at <anonymous> (file:///Users/igl/Library/Caches/deno/npm/registry.npmjs.org/@dagger.io/dagger/0.12.6/dist/index.js:3:10)
```

graphql-tag is a ~~freak~~ "hybrid" package. if i slightly modify their package.json in my deno cache and change the "main" field to the ESM export, i can start dagger with deno:

```sh
$ deno run -A ./deploy.ts
i am running on Deno.version: { deno: "1.46.1", v8: "12.9.202.2-rusty", typescript: "5.5.2" }
Hello from Dagger and Node v16.20.2
```

**Comment by igl** [comment]
Reported the issue with graphql-tag to deno https://github.com/denoland/deno/issues/25311

**Comment by beingminimal** [comment]
> Reported the issue with graphql-tag to deno [denoland/deno#25311](https://github.com/denoland/deno/issues/25311)

To fix this dagger team needs to release typescript packages as ESM modules?
If yes then should i create a separate issue for this?
Awaiting your response. Thanks

**Comment by TomChv** [comment]
Hey, I think we already release the TS package as an ESM modules. But to create dagger module with Deno, we'll need it integrated inside our runtime and also allow the TS SDK to be locally installed.
I'll see if we can use https://github.com/garronej/denoify to convert the Node TS SDK to Deno, it might work

**Comment by shykes** [comment]
Now that we have a SDK called "typescript" it would probably be confusing to have one called "deno". That would essentially be 2 competing Typescript SDKs. Wouldn't it be better to add deno support to the same sdk?

Thoughts @TomChv @helderco @beingminimal ?

**Comment by TomChv** [comment]
> Now that we have a SDK called "typescript" it would probably be confusing to have one called "deno". That would essentially be 2 competing Typescript SDKs. Wouldn't it be better to add deno support to the same sdk?

As we disccused on Discord: https://discord.com/channels/707636530424053791/1157515934597136404/1285595642936295474, I couldn't make denoify to correctly convert our SDK to Deno code. And since https://github.com/denoland/deno/issues/18474 is still open, it's not possible to import local node module to deno code.

So we're pretty much stuck, we cannot use the current TS SDK inside Deno so the solution is to create a different a native deno SDK.

But we cannot mixup-sources, either we have a specific deno SDK directory or, we change the current `typescript`dir to have `node-bun` & `deno` with 2 different codebases, and we can combine them in the TS runtime.

We can in fact keep only 1 runtime, but the SDK/generator will be different.

Also because this requires to make a new codebase, it also means maintaing that codebase, that's why I suggested to make a community Deno SDK, so it's not an official one but make our users happy and indeed give us time to find a way to merge them.

**Comment by shykes** [comment]
@TomChv what about an option 4 where it's 100% the same SDK, but the SDK is configurable to use Node or Deno? This would require a system for passing configuration to SDKs, but that is something that is desirable anyway IMO, and being discussed already. WDYT @helderco ?

**Comment by TomChv** [comment]
> Can we go reverse? Can we release js/ts sdk as package on jsr.io and it will be supported in all runtimes like deno, node & bun?

No, as I already explained, the TS SDK is bundled inside the dagger engine so we cannot do such thing.

**Comment by TomChv** [comment]
> @TomChv what about an option 4 where it's 100% the same SDK, but the SDK is configurable to use Node or Deno? This would require a system for passing configuration to SDKs, but that is something that is desirable anyway IMO, and being discussed already. WDYT @helderco ?

I don't understand how this would solve the original issue that is Deno needs a compatible codebase to run.

**Comment by helderco** [comment]
> @TomChv what about an option 4 where it's 100% the same SDK, but the SDK is configurable to use Node or Deno? This would require a system for passing configuration to SDKs, but that is something that is desirable anyway IMO, and being discussed already. WDYT @helderco ?

The TypeScript SDK is already configurable (same as Python), but the configuration is used for setting up the runtime container. I believe the problem here is in the SDK's client library itself, but I'd like to understand more clearly what's the incompatibility. Can you clarify, and perhaps give a few examples @TomChv?

**Comment by helderco** [comment]
> > Can we go reverse? Can we release js/ts sdk as package on jsr.io and it will be supported in all runtimes like deno, node & bun?
> 
> No, as I already explained, the TS SDK is bundled inside the dagger engine so we cannot do such thing.

Don't rule this out, I can probably think of a few ways to make it happen. Just need to understand how they're incompatible and what would be necessary to make them compatible. We'll think about bundling later.

**Comment by helderco** [comment]
Are there examples of a shared codebase that runs on both node and deno?

**Comment by beingminimal** [comment]
> Are there examples of a shared codebase that runs on both node and deno?

You can check which package is compatible with which runtime on www.jsr.io and also please check this docs it may help.
https://jsr.io/docs/with/node

**Comment by TomChv** [comment]
> The TypeScript SDK is already configurable (same as Python), but the configuration is used for setting up the runtime container (you can already choose to use Bun, for example). I believe the problem here is in the SDK's client library itself, but I'd like to understand more clearly what's the incompatibility. Can you clarify, and perhaps give a few examples @TomChv?

The incompabitility issue is due to the specific Deno import syntax.

Deno has no Node module system but instead load everything like Go would.

```
import express from "npm:express@4.18.2"; // For package published to NPM
import { assert, assertEquals } from "jsr:@std/assert"; // For package published to JSR
```

While TypeScript has `node_modules` and the format you already know.

Appart from that, the codebase should be compatible since we can use npm package in deno... except for native runtime call like `process` who needs to be adapted.

So 2 major things:
- Import syntax that is not compatible between Bun/Node & Deno
- Native library that needs to be adapted.

> Many packages from [www.jsr.io](http://www.jsr.io/) will have a codebase like running on deno and node both.
>
> You can check which package is compatible with which runtime on [www.jsr.io](http://www.jsr.io/) and also Please check this docs it may help.
[jsr.io/docs/with/node](https://jsr.io/docs/with/node)

Yes but they're published to a remote registry, is there a way to import a local JSR package? I didn't find a way from their doc.

**Comment by beingminimal** [comment]
@TomChv Just to add more in context now node supports typescript natively.
https://nodejs.org/en/learn/typescript/run-natively

**Comment by TomChv** [comment]
> @TomChv Just to add more in context now node supports typescript natively.
[nodejs.org/en/learn/typescript/run-natively](https://nodejs.org/en/learn/typescript/run-natively)

Yeah, I'm not sure if this help us though with our current issue

**Comment by beingminimal** [comment]
> > @TomChv Just to add more in context now node supports typescript natively.
> > [nodejs.org/en/learn/typescript/run-natively](https://nodejs.org/en/learn/typescript/run-natively)
> 
> Yeah, I'm not sure if this help us though with our current issue

Inside Typescript SDK code, we are importing .js extension files in .ts files. Instead of transpiling it to js and using it, we can use [nodejs.org/en/learn/typescript/run-natively](https://nodejs.org/en/learn/typescript/run-natively). And Deno supports ts file natively so this can be useful in both ways.

**Comment by TomChv** [comment]
> Inside Typescript SDK code, we are importing .js extension files in .ts files. Instead of transpiling it to js and using it, we can use [nodejs.org/en/learn/typescript/run-natively](https://nodejs.org/en/learn/typescript/run-natively). And Deno supports ts file natively so this can be useful in both ways.

Interesting, however, we're always using the LTS node version, and this is supported in 22.6.0, so I'm not sure about the dependency bump for that.

**Comment by beingminimal** [comment]
> > Inside Typescript SDK code, we are importing .js extension files in .ts files. Instead of transpiling it to js and using it, we can use [nodejs.org/en/learn/typescript/run-natively](https://nodejs.org/en/learn/typescript/run-natively). And Deno supports ts file natively so this can be useful in both ways.
> 
> Interesting, however, we're always using the LTS node version, and this is supported in 22.6.0, so I'm not sure about the dependency bump for that.

Yup! experimentation required.

**Comment by TomChv** [comment]
> Yup! experimentation required.

We gonna wait for october, this is when node 22 will be bump to lts :D

**Comment by helderco** [comment]
> We gonna wait for october, this is when node 22 will be bump to lts :D

That's just release, doesn't mean it can't be experimented and worked on in the meantime. Pending other priorities, of course, unless @beingminimal can help.

**Comment by helderco** [comment]
> The incompabitility issue is due to the specific Deno import syntax.
> 
> Deno has no Node module system but instead load everything like Go would.

I realize that, but I think I also saw compatibility support where you can configure the Deno imports in a separate file, basically mapping the node imports to ones Deno can understand.

**Comment by TomChv** [comment]
Yup, I did a small PoC where you can run deno and tsx with just a few config from `deno.json` and `package.json`:

https://github.com/TomChv/deno-node-compat-test, also it seems tsx supports `.ts` extension so we might be able to introduce it soon enough

**Comment by TomChv** [comment]
Ok did some checks and we hit an issue when trying to run tests (see: https://github.com/TomChv/dagger/pull/347)

I removed the `loader` option to only use it when running with node, it works but I hit another issue due to missing node internal function: https://github.com/denoland/deno/pull/25262

I see that there's a [fix](https://github.com/denoland/deno/pull/25262) but it's not released yet, I'll try with deno canary version if I can make it work

**Comment by TomChv** [comment]
I found a way to get the test running, only 2 are breaking:

```
Establishing connection to Engine... 1   : connect
      2) Should download and unpack the CLI binary automatically


  74 passing (1m)
  2 pending
  2 failing
```

This one is super weird, I checked the type and it's the only test where the type doesn't match even if the setup is exactly the same. I'll play with that to see what happens.

```
1) TypeScript SDK api
       Pass a client with an explicit ID as a parameter:
     TypeError: The "nodePath" option must be a string or a file URL: /Users/tomchauveau/.deno/bin/deno.
      at safeNormalizeFileUrl 
```

Second is a timeout with the engine provisionning. This one isn't important since we don't need that anymore.

```
 2) TypeScript sdk Connect
       Automatic Provisioned CLI Binary
         Should download and unpack the CLI binary automatically:
     Error: Timeout of 30000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/tomchauveau/Documents/github.com/dagger/dagger/sdk/typescript/test/connect.spec.ts)
      at cb (ext:deno_node/internal/timers.mjs:64:31)
      at callback (ext:deno_web/02_timers.js:68:7)
      at eventLoopTick (ext:core/01_core.js:210:13)
```

**Comment by TomChv** [comment]
What's missing to support deno:
- Decoupling provisionning because it's not supported well in Deno.
- Make the code compliant with every runtime + JS (remove .js extension file).

I did a very simple test and it seems we can make it work if we remove extensions:

```ts
// src/fn.ts
export function fn() {
  console.log("foo")
}
```

```typescript
// src/main.ts
import { fn } from "./fn"

function hello() {
  fn()
}

hello()
```

```shell
✗ npx tsc

# Javascript
✗ node dist/main.js                        
foo

# Typescript
✗ tsx src/main.ts                          
foo

# Deno (needs this flag to make it work)
✗ deno run --unstable-sloppy-imports src/main.ts
foo

# Bun
✗ bun run src/main.ts                           
foo
```

If I do not put `--unstable-sloppy-imports`, I get the following error:

```
✗ deno run src/main.ts                          
error: Module not found "file:///Users/tomchauveau/Documents/github.com/quartz-technology/playground/typescript_compat_check/src/fn". Maybe add a '.ts' extension or run with --unstable-sloppy-imports
    at file:///Users/tomchauveau/Documents/github.com/quartz-technology/playground/typescript_compat_check/src/main.ts:1:20
```

**Comment by igl** [comment]
JavaScript modules (ESM) generally require file extensions in import statements. Node (in cjs mode) just can't.
Instead of telling deno to be not-esm-compliant with the --sloppy-imports flag, you can tell node to be esm-compliant.

- Both runtimes do not require the extensions when you import from the node_modules/ folders.
- Deno allows cjs code only when code is inside node_modules/ but not in your own codebase.
- [--experimental-default-typetype](https://nodejs.org/api/cli.html#--experimental-default-typetype) disable the cjs-default
- [--experimental-transform-types](https://nodejs.org/api/cli.html#--experimental-transform-types) to run .ts files

node is adopting many deno features in general right now (test runner, access-control-flags...)

**Comment by TomChv** [comment]
> JavaScript modules (ESM) generally require file extensions in import statements. Node (in cjs mode) just can't.
Instead of telling deno to be not-esm-compliant with the --sloppy-imports flag, you can tell node to be esm-compliant.

However right now, our TS code is using ESM and all our example in the doc etc are also in EMS, wouldn't it break the code and the library to switch to commonJS? There's also some libraries we used that dropped commonJS (https://github.com/dagger/dagger/pull/8237#issue-2487978342) , so it's not a possible solution I think, except if you know if we can make it work somehow?

**Comment by rawkode** [comment]
> Decoupling provisionning because it's not supported well in Deno.

What do you mean by this?

**Comment by beingminimal** [comment]
@shykes @helderco @TomChv 

PLEASE HELP!

[Running TypeScript Natively](https://nodejs.org/en/learn/typescript/run-natively#running-typescript-natively) in node is now supported.

And so we can have a single package on jsr.io only for node and deno both.

We're eager to use Deno with the Dagger engine. Adding experimental Deno support, perhaps after publishing a package on jsr.io, would be a huge help and unblock several of our projects. This is impacting our current work. We would be very grateful if you could consider adding this support.

Could you please advise on the best approach to achieve this? For example, would integrating Deno directly into the Dagger engine and then publishing a package on jsr.io be the recommended path?

Thanks
